osu.Desktop/Windows/WindowsAssociationManager.cs

```diff
@@ -0,0 +1,273 @@
+                    return s.ToString();
+
+                var b = localisation.GetLocalisedBindableString(s);
+                b.UnbindAll();
+                return b.Value;
+            }
+        }
+
+        #region Native interop
+
+        [DllImport("Shell32.dll")]
+        private static extern void SHChangeNotify(EventId wEventId, Flags uFlags, IntPtr dwItem1, IntPtr dwItem2);
+
+        private enum EventId
+        {
+            /// <summary>
+            /// A file type association has changed. <see cref="Flags.SHCNF_IDLIST"/> must be specified in the uFlags parameter.
+            /// dwItem1 and dwItem2 are not used and must be <see cref="IntPtr.Zero"/>. This event should also be sent for registered protocols.
+            /// </summary>
+            SHCNE_ASSOCCHANGED = 0x08000000
+        }
+
+        private enum Flags : uint
+        {
+            SHCNF_IDLIST = 0x0000
+        }
+
+        #endregion
+
+        private record FileAssociation(string Extension, LocalisableString Description, string IconPath)
+        {
+            private string programId => $@"{PROGRAM_ID_PREFIX}{Extension}";
+
+            /// <summary>
+            /// Installs a file extenstion association in accordance with https://learn.microsoft.com/en-us/windows/win32/com/-progid--key
+            /// </summary>
+            public void Install(RegistryKey classes, string exePath)
+            {
+                // register a program id for the given extension
+                using (var programKey = classes.CreateSubKey(programId))
+                {
+                    using (var defaultIconKey = programKey.CreateSubKey(DEFAULT_ICON))
+                        defaultIconKey.SetValue(null, IconPath);
+
+                    using (var openCommandKey = programKey.CreateSubKey(SHELL_OPEN_COMMAND))
+                        openCommandKey.SetValue(null, $@"""{exePath}"" ""%1""");
+                }
+
+                using (var extensionKey = classes.CreateSubKey(Extension))
+                {
+                    // set ourselves as the default program
+                    extensionKey.SetValue(null, programId);
+
+                    // add to the open with dialog
+                    // https://learn.microsoft.com/en-us/windows/win32/shell/how-to-include-an-application-on-the-open-with-dialog-box
+                    using (var openWithKey = extensionKey.CreateSubKey(@"OpenWithProgIds"))
+                        openWithKey.SetValue(programId, string.Empty);
+                }
+            }
+
+            public void UpdateDescription(RegistryKey classes, string description)
+            {
+                using (var programKey = classes.OpenSubKey(programId, true))
+                    programKey?.SetValue(null, description);
+            }
+
+            public void Uninstall(RegistryKey classes)
+            {
+                // importantly, we don't delete the default program entry because some other program could have taken it.
```

Comment:
Doesn't this mean there'll be a hanging entry there? Is windows going to handle that fine?

Wouldn't it be better to just check if the key can be safely uninstalled (if it has the expected value then it can be removed)?

Reply:
> Doesn't this mean there'll potentially be a hanging entry there to a nonexistent executable?

Hanging entry to a nonexistent ProgID. [MS recommends to just leave it like that](https://learn.microsoft.com/en-us/windows/win32/shell/fa-file-types#deleting-registry-information-during-uninstallation) (I'll add this link to the inline comment).

> Is windows going to handle that fine?

The only visible issue is that the associated file types will be shown as "osu.File.osz" in explorer. Whereas windows by default would show "OSZ file".

![image](https://github.com/ppy/osu/assets/16479013/f360a46d-323a-4826-84c2-8dcf2f728940)

![image](https://github.com/ppy/osu/assets/16479013/5f1eb13b-3f69-427d-8aa5-acae3c1494a9)

> Wouldn't it be better to just check if the key can be safely uninstalled (if it has the expected value then it can be removed)?

Yeah, better to remove so it doesn't show "osu.File.*" to the user.